File: chapters/src/main/java/cn/rubintry/chapters/viewmodel/LoginAndRegisterViewModel.java
package cn.rubintry.chapters.viewmodel;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import cn.rubintry.common.model.LoginModel;
import cn.rubintry.common.model.RegisterModel;
import cn.rubintry.common.model.BaseModel;
import cn.rubintry.common.utils.http.OkHttpUtils;
import cn.rubintry.common.utils.http.ResponseCallback;

/**
 * @author logcat
 */
public class LoginAndRegisterViewModel extends ViewModel {
    private MutableLiveData<BaseModel<LoginModel>> userData;
    private MutableLiveData<BaseModel<RegisterModel>> registerData;


    /**
     * 监听登录数据
     * @param userName
     * @param password
     * @return
     */
    public MutableLiveData<BaseModel<LoginModel>> login(String userName , String password) {
        if (userData == null) {
            userData = new MutableLiveData<>();
        }
        requestLogin(userName , password);
        return userData;
    }


    /**
     * 发起登录请求
     * @param userName
     * @param password
     */
    private void requestLogin(String userName , String password) {
        OkHttpUtils.getInstance()
                .post("/user/login")
                .addParameters("username" , userName)
                .addParameters("password" , password)
                .request(new ResponseCallback<BaseModel<LoginModel>>() {
                    @Override
                    protected void onSuccess(BaseModel<LoginModel> loginModelBaseModel) {
                        userData.postValue(loginModelBaseModel);
                    }

                    @Override
                    protected void onFailed(Exception e, BaseModel<LoginModel> loginModelBaseModel) {
                        userData.postValue(loginModelBaseModel);
                    }

                });
    }


    /**
     * 监听注册账号
     * @param username
     * @param password
     * @param rePassword
     * @return
     */
    public MutableLiveData<BaseModel<RegisterModel>> register(String username , String password , String rePassword) {

        if(registerData == null){
            registerData = new MutableLiveData<>();
        }

        requestRegister(username , password , rePassword);

        return registerData;
    }


    /**
     * 发起注册请求
     * @param username
     * @param password
     * @param rePassword
     */
    private void requestRegister(String username , String password , String rePassword) {
        OkHttpUtils.getInstance()
                .post("/user/register")
                .addParameters("username" , username)
                .addParameters("password" , password)
                .addParameters("repassword" , rePassword)
                .request(new ResponseCallback<BaseModel<RegisterModel>>() {
                    @Override
                    protected void onSuccess(BaseModel<RegisterModel> registerModelBaseModel) {
                        registerData.postValue(registerModelBaseModel);
                    }

                    @Override
                    protected void onFailed(Exception e, BaseModel<RegisterModel> registerModelBaseModel) {
                        registerData.postValue(registerModelBaseModel);
                    }
                });
    }
}

File: chapters/src/main/java/cn/rubintry/chapters/viewmodel/LoginAndRegisterViewModel.kt
package cn.rubintry.chapters.viewmodel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import cn.rubintry.common.model.BaseModel
import cn.rubintry.common.model.LoginModel
import cn.rubintry.common.model.RegisterModel
import cn.rubintry.common.utils.http.OkHttpUtils
import cn.rubintry.common.utils.http.ResponseCallback

/**
 * @author logcat
 */
class LoginAndRegisterViewModel : ViewModel() {
    private var userData: MutableLiveData<BaseModel<LoginModel>>? = null
    private var registerData: MutableLiveData<BaseModel<RegisterModel>>? = null

    /**
     * 监听登录数据
     * @param userName
     * @param password
     * @return
     */
    fun login(
        userName: String,
        password: String
    ): MutableLiveData<BaseModel<LoginModel>> {
        if (userData == null) {
            userData = MutableLiveData()
        }
        requestLogin(userName, password)
        return userData as MutableLiveData<BaseModel<LoginModel>>
    }

    /**
     * 发起登录请求
     * @param userName
     * @param password
     */
    private fun requestLogin(userName: String, password: String) {
        OkHttpUtils.instance
            ?.post("/user/login")
            ?.addParameters("username", userName)
            ?.addParameters("password", password)
            ?.request(object : ResponseCallback<BaseModel<LoginModel>>(){
                override fun onSuccess(t: BaseModel<LoginModel>) {
                    userData?.postValue(t)
                }

                override fun onFailed(e: Exception?, t: BaseModel<LoginModel>?) {
                    userData?.postValue(t)
                }
            })
    }

    /**
     * 监听注册账号
     * @param username
     * @param password
     * @param rePassword
     * @return
     */
    fun register(
        username: String,
        password: String,
        rePassword: String
    ): MutableLiveData<BaseModel<RegisterModel>> {
        if (registerData == null) {
            registerData = MutableLiveData()
        }
        requestRegister(username, password, rePassword)
        return registerData as MutableLiveData<BaseModel<RegisterModel>>
    }

    /**
     * 发起注册请求
     * @param username
     * @param password
     * @param rePassword
     */
    private fun requestRegister(
        username: String,
        password: String,
        rePassword: String
    ) {
        OkHttpUtils.instance
            ?.post("/user/register")
            ?.addParameters("username", username)
            ?.addParameters("password", password)
            ?.addParameters("repassword", rePassword)
            ?.request(object : ResponseCallback<BaseModel<RegisterModel>>(){
                override fun onSuccess(t: BaseModel<RegisterModel>) {
                    registerData?.postValue(t)
                }

                override fun onFailed(e: Exception?, t: BaseModel<RegisterModel>?) {
                    registerData?.postValue(t)
                }

            })
    }
}